### ISSUE TITLE
Blazor project consume all possible memory and server hangs up.

### ISSUE THREAD

**Issue by YordanYanakiev** [dotnet/aspnetcore#29206]
<!--

More information on our issue management policies can be found here: https://aka.ms/aspnet/issue-policies
### Describe the bug
Blazor project does not release the allocated memory in Application pool, and the application pool fastly grows ginormous, practically until all possible memory is allocated, and then the server start rejecting new connections, and hand up.
The problem grows absurdly fast if I try to use 3rd party components like Syncfusion's. It cost about a day to consume about 100 GB RAM, with just 300 users, and then crash the server, and freeze the rest of the applications even.

### To Reproduce
Just put some code alike:
```
@for( int c = 0; c < 2000; c++ )
{
    <img src="@( @"1.jpg?" + c.ToString() )"  /><br />
}
```
**Run it, and start connecting to the site from different browsers and computers.**

### Exceptions (if any)
No practical exception during the hungry Application pool

### Further technical details
- The issue is observed on Windows 8.1 (developer pc in debug mode), Windows 10 (developer pc in debug and production mode), Windows Server 2016 (production server with the release )
- ASP.NET Core version : affected on both 3.1 and 5.0
- Include the output of `dotnet --info` :
C:\Users\Sun>dotnet --info

> .NET SDK (reflecting any global.json):
>  Version:   5.0.200-preview.20601.7
>  Commit:    b3b934bbf2
> 
> Runtime Environment:
>  OS Name:     Windows
>  OS Version:  6.3.9600
>  OS Platform: Windows
>  RID:         win81-x64
>  Base Path:   C:\Program Files\dotnet\sdk\5.0.200-preview.20601.7\
> 
> Host (useful for support):
>   Version: 5.0.1
>   Commit:  b02e13abab
> 
> .NET SDKs installed:
>   1.1.0 [C:\Program Files\dotnet\sdk]
>   2.0.3 [C:\Program Files\dotnet\sdk]
>   2.1.2 [C:\Program Files\dotnet\sdk]
>   2.1.4 [C:\Program Files\dotnet\sdk]
>   2.1.100 [C:\Program Files\dotnet\sdk]
>   2.1.202 [C:\Program Files\dotnet\sdk]
>   2.1.602 [C:\Program Files\dotnet\sdk]
>   2.1.801 [C:\Program Files\dotnet\sdk]
>   2.2.202 [C:\Program Files\dotnet\sdk]
>   2.2.401 [C:\Program Files\dotnet\sdk]
>   3.0.100 [C:\Program Files\dotnet\sdk]
>   3.1.100-preview1-014459 [C:\Program Files\dotnet\sdk]
>   3.1.201 [C:\Program Files\dotnet\sdk]
>   3.1.300-preview-015115 [C:\Program Files\dotnet\sdk]
>   3.1.300 [C:\Program Files\dotnet\sdk]
>   5.0.101 [C:\Program Files\dotnet\sdk]
>   5.0.200-preview.20601.7 [C:\Program Files\dotnet\sdk]
> 
> .NET runtimes installed:
>   Microsoft.AspNetCore.All 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
>   Microsoft.AspNetCore.All 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
>   Microsoft.AspNetCore.All 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
>   Microsoft.AspNetCore.All 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
>   Microsoft.AspNetCore.All 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
>   Microsoft.AspNetCore.All 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
>   Microsoft.AspNetCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 3.1.0-preview1.19508.20 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 3.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.NETCore.App 1.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 1.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 2.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 2.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 3.1.0-preview1.19506.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 3.1.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.WindowsDesktop.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 3.1.0-preview1.19506.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 3.1.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
> 

- The IDE (VS / VS Code/ VS4Mac) you're running on, and its version: 2019 preview / Version 16.9.0 Preview 2.0
.

**Comment by YordanYanakiev** [comment]
### Describe the bug
Blazor project does not release the allocated memory in Application pool, and the application pool fastly grows ginormous, practically until all possible memory is allocated, and then the server start rejecting new connections, and hand up.

### To Reproduce
Just put some code alike:
```
@for( int c = 0; c < 2000; c++ )
{
    <img src="@( @"1.jpg?" + c.ToString() )"  /><br />
}
```
**Run it, and start connecting to the site from different browsers and computers.**

### Exceptions (if any)
No practical exception during the hungry Application pool

### Further technical details
- The issue is observed on Windows 8.1 (developer pc in debug mode), Windows 10 (developer pc in debug and production mode), Windows Server 2016 (production server with the release )
- ASP.NET Core version : affected on both 3.1 and 5.0
- Include the output of `dotnet --info` :
C:\Users\Sun>dotnet --info

> .NET SDK (reflecting any global.json):
>  Version:   5.0.200-preview.20601.7
>  Commit:    b3b934bbf2
> 
> Runtime Environment:
>  OS Name:     Windows
>  OS Version:  6.3.9600
>  OS Platform: Windows
>  RID:         win81-x64
>  Base Path:   C:\Program Files\dotnet\sdk\5.0.200-preview.20601.7\
> 
> Host (useful for support):
>   Version: 5.0.1
>   Commit:  b02e13abab
> 
> .NET SDKs installed:
>   1.1.0 [C:\Program Files\dotnet\sdk]
>   2.0.3 [C:\Program Files\dotnet\sdk]
>   2.1.2 [C:\Program Files\dotnet\sdk]
>   2.1.4 [C:\Program Files\dotnet\sdk]
>   2.1.100 [C:\Program Files\dotnet\sdk]
>   2.1.202 [C:\Program Files\dotnet\sdk]
>   2.1.602 [C:\Program Files\dotnet\sdk]
>   2.1.801 [C:\Program Files\dotnet\sdk]
>   2.2.202 [C:\Program Files\dotnet\sdk]
>   2.2.401 [C:\Program Files\dotnet\sdk]
>   3.0.100 [C:\Program Files\dotnet\sdk]
>   3.1.100-preview1-014459 [C:\Program Files\dotnet\sdk]
>   3.1.201 [C:\Program Files\dotnet\sdk]
>   3.1.300-preview-015115 [C:\Program Files\dotnet\sdk]
>   3.1.300 [C:\Program Files\dotnet\sdk]
>   5.0.101 [C:\Program Files\dotnet\sdk]
>   5.0.200-preview.20601.7 [C:\Program Files\dotnet\sdk]
> 
> .NET runtimes installed:
>   Microsoft.AspNetCore.All 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
>   Microsoft.AspNetCore.All 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
>   Microsoft.AspNetCore.All 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
>   Microsoft.AspNetCore.All 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
>   Microsoft.AspNetCore.All 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
>   Microsoft.AspNetCore.All 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
>   Microsoft.AspNetCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 3.1.0-preview1.19508.20 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 3.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.AspNetCore.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
>   Microsoft.NETCore.App 1.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 1.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 2.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 2.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 3.1.0-preview1.19506.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 3.1.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.NETCore.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>   Microsoft.WindowsDesktop.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 3.1.0-preview1.19506.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 3.1.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 3.1.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
>   Microsoft.WindowsDesktop.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
> 

- The IDE (VS / VS Code/ VS4Mac) you're running on, and its version: 2019 preview / Version 16.9.0 Preview 2.0

**Comment by javiercn** [comment]
@YordanYanakiev thanks for contacting us.

This is not a Blazor issue but a developer error. We don't recommend allocating large strings (like the one you are creating) for serving images as data URLs. 

By doing that you are creating large strings that are alive while the user session is alive. We recommend that instead you serve the images from an API endpoint.

We have a separate issue to provide a component that makes this scenario easier in the future.

**Comment by YordanYanakiev** [comment]
I am just pointing a fact trough easily reproducing test.
And this is actually extremely small issue.
Using Syncfusion's components the things getting uglier in a mater of seconds with large data providers.
Therefore I assume that something is not as it should be with Blazor disregarding what is recommended or not, since after all none application could be just 3 rows big.

**Comment by javiercn** [comment]
@YordanYanakiev 

Just to confirm, the images are base64 encoded as a data url, is that the case?

_Using Syncfusion's components the things getting uglier in a mater of seconds with large data providers._
Therefore I assume that something is not as it should be with Blazor disregarding what is recommended or not, since after all none application could be just 3 rows big.
Can you point out what component is causing the issue? @danroth27 maybe we can ping them and offer alternatives.

_And under "user session".. I am not sure which one is pointing - the session which is opened on the server by IIS user, or the session which is opened by the remote user itself, because if this is the case - then after the user disconnect nothing is cleaned by the Garbage collector, and the Application pool only keeps expanding even if everyone user disconnects, and in a mater of minutes everything goes bad._

I'm referring to the user circuit, the circuit stays in a disconnected state for a couple of minutes before it gets disposed and eligible for GC.

_And if the strings is the issue then how I am supposed to write applications which is supposed to visualize large amounts of data trough data models?_

Like any other app, when you build you app you have a high impact on the performance of the application, you should be able to calculate how much memory each user/page requires and how you build the page will have an impact on it. If your model includes images and you decide to base64 encode them as data urls, then you are potentially creating large strings in memory. Assuming 1mb per image and 20000 rows without virtualization this is about ~20GB per user.

In general you shouldn't have problems about model size in Blazor applications unless you are dealing with images or other media sources in the way I described. In the end, if your app is large, complex, and needs to load a lot of data you'll have to do work to make it perform efficiently. This is not unique to Blazor but to every UI application framework.

As a good rule of thumb, the amount of data a user can see is limited to the space in their monitor, so you can always optimize your app to only load in memory the minimal data that fits on the screen at a time and the necessary components to render it.

_Aside what is the difference by serving from API endpoint ?_

Based on my assumption about the url being base64url encoded data url strings, you are creating a large payload on the server that needs to be sent to the client and due to it being an attribute the server needs to maintain a reference to it. By serving the contents from an endpoint, the actual image doesn't become part of the render tree and blazor doesn't need to hold on to it for the lifetime of the circuit for diffing purposes and render updates.

_On the actually application I am supposed to show to the users not tons of data which is visualized directly without any kind of paging. If Blazor is not capable of such basic things it should be coming with a extreme large warning about it._

Blazor is absolutely capable of it, see [here](https://devblogs.microsoft.com/aspnet/blazor-server-in-net-core-3-0-scenarios-and-performance/) for details about scalability. That said, when you are dealing with large amounts of data you'll have to balance whether you want to optimize your app or pay for more servers/memory.

**Comment by YordanYanakiev** [comment]
1. The circuits is long gone ( 2+ hours ) but the memory is not returned back.
2. As I have mentioned - the example i gave is absolutely only for showing the issue. 
My code is nothing alike - it's by the standard recommendations with data models, to a data provider, attached to the components, with almost none images, - probably about 2-3 images which is repeantingly presented with size 32x32 pixels jpeg.
Yet the result is the same. Actually even worse.
3. Currently the number of users is 60, and the problem is on.

**Comment by mkArtakMSFT** [comment]
Thank you for filing this issue. In order for us to investigate this issue, please provide a minimalistic repro project (ideally a GitHub repo) that illustrates the problem.

**Comment by YordanYanakiev** [comment]
I've waited for this mater to get an results from Syncfusion before return to this topic as well, but they turned it into private request where the things start rotating again into endless loop between me and them.
Here the topic with example which present the bug.
https://www.syncfusion.com/forums/161102/extreme-memory-leaks?fbclid=IwAR0iCSyYeKtvuLCvZlLnmZIxT60H_nzFicMyeB77x-zBILLgYVR1QDEF4lM
Here and the screenshot i've made with explanations on it.
![download](https://user-images.githubusercontent.com/6534199/104440550-2d1b7c80-559b-11eb-98ff-0cf40a7c9af0.png)
There is a further development. (which could be useful maybe)
- If I set a very "attacking" Garbage Collector, like a timer set on 3-5 seconds which is requiring GC to act - the memory problem start to looks better, anyway at some point an issue with User Interface components start to occur just because of this.
